src/read/magic_finder.rs

```diff
@@ -0,0 +1,191 @@
+use std::io::{Read, Seek, SeekFrom};
+
+use memchr::memmem::FinderRev;
+
+use crate::result::ZipResult;
+
+/// A utility for finding magic symbols from the end of a seekable reader.
+///
+/// Can be repurposed to recycle the internal buffer.
+pub struct MagicFinder<'a> {
+    buffer: Box<[u8]>,
+    pub(self) finder: FinderRev<'a>,
+    cursor: u64,
+    mid_buffer_offset: Option<usize>,
+    bounds: (u64, u64),
+}
+
+impl<'a> MagicFinder<'a> {
+    /// Create a new magic bytes finder to look within specific bounds.
+    pub fn new(magic_bytes: &'a [u8], bounds: (u64, u64)) -> Self {
+        const BUFFER_SIZE: usize = 2048;
+
+        // Smaller buffer size would be unable to locate bytes.
+        // Equal buffer size would stall (the window could not be moved).
+        debug_assert!(BUFFER_SIZE > magic_bytes.len());
+
+        Self {
+            buffer: vec![0; BUFFER_SIZE].into_boxed_slice(),
+            finder: FinderRev::new(magic_bytes),
+            cursor: bounds
+                .1
+                .saturating_sub(BUFFER_SIZE as u64)
+                .clamp(bounds.0, bounds.1),
+            mid_buffer_offset: None,
+            bounds,
+        }
+    }
+
+    /// Repurpose the finder for different bytes or bounds.
+    pub fn repurpose(&mut self, magic_bytes: &'a [u8], bounds: (u64, u64)) -> &mut Self {
+        debug_assert!(self.buffer.len() > magic_bytes.len());
+
+        self.finder = FinderRev::new(magic_bytes);
+        self.cursor = bounds
+            .1
+            .saturating_sub(self.buffer.len() as u64)
+            .clamp(bounds.0, bounds.1);
+        self.bounds = bounds;
+
+        // Reset the mid-buffer offset, to invalidate buffer content.
+        self.mid_buffer_offset = None;
+
+        self
+    }
+
+    /// Find the next magic bytes from the end of the reader.
+    ///
+    /// Similar in functionality to a double ended iterator, except
+    /// it propagates errors first and doesn't hold on to the reader
+    /// between items.
+    pub fn next_back<R: Read + Seek>(&mut self, reader: &mut R) -> ZipResult<Option<u64>> {
+        loop {
+            if self.cursor < self.bounds.0 {
+                // The finder is consumed
```

Comment:
Should actually be `self.cursor <= self.bounds.0`, since we set them equal at the end.

Reply:
`self.cusor == self.bounds.0` is a valid state. It is achievable by having data of size `BUFFER_SIZE + (BUFFER_SIZE - MAGIC_SIZE + 1) * n` for some `n`. What we set equal at the end are the bounds, to essentially make it a cusor in an empty region.